Manually Correct Matching Algorithm

emichaud998 opened the pull request:
Changes in this PR:
This PR adds the ability to manually correct the matching algorithm through the use of whitelist and blacklist files. After the linking algorithm has ran and stored all the resulting endpoint to npi organization links, the algorithm then runs a function that uses the whitelist and blacklist files to manually add/delete links from the database. The whitelist file (linkerMatchesWhitelist.json) and the blacklist file (linkerMatchesBlacklist.json) can both be found in the resources/prod_resources directory, and the files are mounted to the endpoint manager docker container among startup. The format for these files can be found in the endpoint manager README file, but for now the files have been left empty. 

TO TEST:
To manually test the whitelist and blacklist functionality, add any endpoint to npi organization links that you want to add/remove from the database to the whitelist/blacklist files found in the resources/prod_resources directory following the format shown in the endpoint manager README. Any link added to the whitelist file will be added to the endpoint_organization table in the database, and any link added to the blacklist file will be removed from the endpoint_organization table in the database.

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: [Lantern 250 JIRA Ticket](https://oncprojectracking.healthit.gov/support/browse/LANTERN-250)
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [ ] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [ ] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.

bmath10 commented on endpointmanager/pkg/endpointmanager/postgresql/npiorganizationstore.go:
```diff
@@ -18,6 +18,7 @@ var deleteNPIOrganizationStatement *sql.Stmt
 var linkNPIOrganizationToFHIREndpointStatement *sql.Stmt
 var getNPIOrganizationFHIREndpointLinkStatement *sql.Stmt
 var updateNPIOrganizationFHIREndpointLinkLink *sql.Stmt
+var deleteNPIOrganizationFHIREndpointLinkLink *sql.Stmt
```
It looks like this and the update statement ending with LinkLink is an error?

emichaud998 commented on e2e/integration_tests/data_flow_test.go:
```diff
@@ -205,7 +205,7 @@ func Test_EndpointLinksAreAvailable(t *testing.T) {
 
 	ctx := context.Background()
-	endpointlinker.LinkAllOrgsAndEndpoints(ctx, store, false)
+	endpointlinker.LinkAllOrgsAndEndpoints(ctx, store, "", "", false)
```
I did this because there is nothing in them, and since the whitelist/blacklist files could contain anything and are subject to change, I figured for the data_flow test it would be better not to use them, especially since I have testing for it elsewhere. If my intuition was wrong I can fix this, just let me know!

emichaud998 replied:
Added some testing by creating a fake whitelist and blacklist file in the test data for the dataflow test

bmath10 replied:
oh cool, the test data was a good idea!

emichaud998 commented on endpointmanager/pkg/endpointlinker/endpointlinker_integration_test.go:
```diff
@@ -151,6 +152,78 @@ func Test_addMatch(t *testing.T) {
+	th.Assert(t, sEpURL == ep4URL, fmt.Sprintf("expected stored url '%s' to be the same as the url that was stored from whitelist '%s'.", sEpURL, ep4URL))
+	th.Assert(t, sConfidence == 1.000, fmt.Sprintf("expected stored confidence 1.000, got '%f'.", sConfidence))
+
+	sNpiID, sEpURL, sConfidence, err = store.GetNPIOrganizationFHIREndpointLink(ctx, npiID1, ep3.URL)
```
Based on how I understood the task, since the link between the 3rd endpoint and Organization 3 is put in the blacklist file, this link should be removed from the endpoint_organization table, so this test checks to see if the link between the 3rd endpoint and Organization 3 was removed correctly

bmath10 replied:
ohhh i misread this, I understand now, makes sense!

bmath10 commented:
I tried running `make populatedb` with nothing in the whitelist/blacklist files and i got the error `signal: killed`. i assume it has something to do with reading an empty file (either that or something is wrong with my setup). it seems like it would be a good idea to check whether a file is empty or not (if that's possible) while we currently don't have any real data to run with?

emichaud998 commented:
Bri, I must have deleted the line to check for that when I was making changes, I will fix that now

bmath10 commented:
Two things I noticed while running:
1. If a URL is in the whitelist with a given organization id, it adds it to the endpoint_organization table with no problem, but it doesn't remove any of the other ids that were matched to that URL. My initial thought was that all other matches would be removed if we have a confirmed match; is there a reason why we keep those other matches around?
2. For some reason it seems like if I try to run `make populatedb` twice, the first time it works fine, but the second time it gets to the linker and then seems to time out or something because it sits there for like 30 seconds and then all I get is a `signal: killed` error. Have you seen this/any ideas why this might be happening?

emichaud998 commented:
In regards to your first point, I asked Matt about this, and he said that users can specify additions and deletions, so if they ask for a link to be added, then we can add that to the whitelist, and then we can ask them if they want any of the existing links to be removed and add those to the blacklist.

In regards to your second point, I have not noticed that, but I will take a look at it now!

bmath10 commented:
> In regards to your first point, I asked Matt about this, and he said that users can specify additions and deletions, so if they ask for a link to be added, then we can add that to the whitelist, and then we can ask them if they want any of the existing links to be removed and add those to the blacklist.

Okay that makes sense, thanks for the clarification!

bmath10 commented:
I was just messing around with this again to see if I was still seeing that same problem, and I found out a little more information:
1. Every time I ran `make run` I didn't have a problem with the `endpointlinker` initially. If the `linkMatchesWhitelist` file stayed the same every time I ran `make populatedb` then it worked fine. As soon as I changed `linkMatchesWhitelist` it would break and wouldn't work even if I changed it back to the original file. Stopping everything and running `make run` again seems to fix it, so I'm wondering if it has something to do with the files being mounted as volumes?
2. It seemed to specifically break around the `ioutil.ReadAll(jsonFile)` in the `openLinkerCorrectionFiles` function, or at least when i was using print statements it would get to the point before that call, hang for 30+ seconds, then throw the `signal: killed` error.
If this is just because I'm not supposed to be editing files that are mounted without restarting the container then I guess we don't have to worry about this and I can approve it! If it's something else though, and you're not seeing this error, I guess it's just something weird happening with my computer...

emichaud998 commented:
> I was just messing around with this again to see if I was still seeing that same problem, and I found out a little more information:
> 
> 1. Every time I ran `make run` I didn't have a problem with the `endpointlinker` initially. If the `linkMatchesWhitelist` file stayed the same every time I ran `make populatedb` then it worked fine. As soon as I changed `linkMatchesWhitelist` it would break and wouldn't work even if I changed it back to the original file. Stopping everything and running `make run` again seems to fix it, so I'm wondering if it has something to do with the files being mounted as volumes?
> 2. It seemed to specifically break around the `ioutil.ReadAll(jsonFile)` in the `openLinkerCorrectionFiles` function, or at least when i was using print statements it would get to the point before that call, hang for 30+ seconds, then throw the `signal: killed` error.
> 
> If this is just because I'm not supposed to be editing files that are mounted without restarting the container then I guess we don't have to worry about this and I can approve it! If it's something else though, and you're not seeing this error, I guess it's just something weird happening with my computer...

I also get this same error, but from doing some research, I have found that If you are using some editors, like vim for example and possibly visual studio, when you save the file it does not save the file directly, but it instead creates a new file and copies it into its place. This breaks the bind-mount of the docker container, which is based on inode, which I researched and inode is a data structure in a Unix-style file system that describes a file-system object. Since saving the file, then, is changing the inode, and this is what may be causing that error we are seeing. When the container is restarted, this creates a new inode, which explains why it would work after restarting the container. Therefore I think this is an unavoidable error that can be avoided by just not chaning mounted files without restarting the container, but if you guys find this is not the case, I am happy to talk about it and try to fix the issue!
